tests/test_convolution_1d.py

```diff
@@ -0,0 +1,194 @@
+        raise_exception = trial_counts.shape[1] < basis_matrix.shape[0]
+        if raise_exception:
+            with pytest.raises(ValueError, match="Insufficient trial duration. The number of time points"):
+                utils.convolve_1d_trials(basis_matrix, trial_counts)
+        else:
+            utils.convolve_1d_trials(basis_matrix, trial_counts)
+
+    @pytest.mark.parametrize("basis_matrix", [np.zeros((4, 3))])
+    @pytest.mark.parametrize("trial_counts", [
+        np.zeros((0, 4, 3)),  # invalid
+        np.zeros((1, 40, 0)),  # invalid
+    ])
+    def test_empty_counts(self, basis_matrix, trial_counts):
+        with pytest.raises(ValueError, match="time_series should not contain"):
+            utils.convolve_1d_trials(basis_matrix, trial_counts)
+
+    @pytest.mark.parametrize("basis_matrix", [np.random.normal(size=(4, 3)) for _ in range(2)])
+    @pytest.mark.parametrize("trial_counts", [np.random.normal(size=(2, 10, 3)) for _ in range(2)])
+    def test_valid_convolution_output(self, basis_matrix, trial_counts):
+        numpy_out = np.zeros((trial_counts.shape[0],
+                              trial_counts.shape[1] - basis_matrix.shape[0] + 1,
+                              trial_counts.shape[2],
+                              basis_matrix.shape[1]))
+        for tri_i, trial in enumerate(trial_counts):
+            for neu_k, vec in enumerate(trial.T):
+                for bas_j, basis in enumerate(basis_matrix.T):
+                    numpy_out[tri_i, :, neu_k, bas_j] = np.convolve(vec, basis, mode='valid')
+
+        utils_out = np.asarray(utils.convolve_1d_trials(basis_matrix, trial_counts))
+        assert np.allclose(utils_out, numpy_out, rtol=10**-5, atol=10**-5), "Output of utils.convolve_1d_trials " \
+                                                                            "does not match numpy.convolve in " \
+                                                                            "\"valid\" mode."
+
+
+class TestPadding:
+
+    @pytest.mark.parametrize("filter_type", ["causal", "acausal", "anti-causal"])
+    @pytest.mark.parametrize("iterable", [[np.zeros([1]*n)] * 2 for n in range(1, 6)] +
+                                         [[np.zeros([1, 2, 4]), np.zeros([1, 2, 4])]] +
+                                         [[np.zeros([1, 2, 4]), np.zeros([1, 1, 1, 1])]] +
+                                         [[np.zeros([1, 2, 4, 5]), np.zeros([1, 1, 1, 1, 1])]]
+    )
+    def test_check_dim(self, iterable, filter_type):
+        raise_exception = any(trial.ndim != 3 for trial in iterable)
+        if raise_exception:
+            with pytest.raises(ValueError, match="conv_trials must be an iterable of 3D arrays"):
+                utils.nan_pad_conv(iterable, 3, filter_type)
+        else:
+            utils.nan_pad_conv(iterable, 3, filter_type)
+
+    @pytest.mark.parametrize("filter_type", ["causal", "acausal", "anti-causal", ""])
```

Comment:
probably don't need to test the allowed `filter_types` here, because otherwise this is identical to `test_check_dim`

Reply:
I removed looping on filter_types on the test_check_dim. The difference between the tests was that this one raises an exception